### PR TITLE
Add more callers for table reader. 

### PR DESCRIPTION
**Pull request by HaoyuHuang**
This PR adds more callers for table readers. These information are only used for block cache analysis so that we can know which caller accesses a block. 
1. It renames the BlockCacheLookupCaller to TableReaderCaller as passing the caller from upstream requires changes to table_reader.h and TableReaderCaller is a more appropriate name. 
2. It adds more table reader callers in table/table_reader_caller.h, e.g., kCompactionRefill, kExternalSSTIngestion, and kBuildTable. 

This PR is long as it requires modification of interfaces in table_reader.h, e.g., NewIterator. 

Test plan: make clean && COMPILE_WITH_ASAN=1 make check -j32.

### CODE REVIEW

File: table/table_reader_caller.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+namespace rocksdb {
+// A list of callers for a table reader. It is used to trace the caller that
+// accesses on a block. This is only used for block cache analysis.
+enum TableReaderCaller : char {
```
Comment (riversand963) [inline table/table_reader_caller.h]: Are all these necessary? Do we want to distinguish between 'SSTFileReader', 'repair', 'unit test', etc.?

Comment (HaoyuHuang) [reply table/table_reader_caller.h]: I agree that we may not need to differentiate these callers. 

I originally only want to add compactionRefill and SSTIngestion. As I changed the NewIterator in table reader, I realize many other places touch the NewIterator. So later I put these callers in as they can be differentiated easily.

File: db/builder.cc [hunk 2]
```diff
@@ -221,8 +221,9 @@ Status BuildTable(
           mutable_cf_options.prefix_extractor.get(), nullptr,
           (internal_stats == nullptr) ? nullptr
                                       : internal_stats->GetFileReadHist(0),
-          false /* for_compaction */, nullptr /* arena */,
-          false /* skip_filter */, level));
+          TableReaderCaller::kBuildTable, /*arena=*/nullptr,
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply db/builder.cc]: Sure. I renamed it to kFlush.

File: db/table_cache.cc [hunk 3]
```diff
@@ -203,6 +203,7 @@ InternalIterator* TableCache::NewIterator(
     *table_reader_ptr = nullptr;
   }
   size_t readahead = 0;
+  bool for_compaction = caller == TableReaderCaller::kCompaction;
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply db/table_cache.cc]: No, it is not. TableReaderCaller::kCompactionRefill is only set here. https://github.com/facebook/rocksdb/blob/master/db/compaction/compaction_job.cc#L653
I respect the previous behavior which sets for_compaction to false.

File: table/block_based/block_based_table_reader.cc [hunk 4]
```diff
@@ -2944,7 +2944,7 @@ void BlockBasedTable::MultiGet(const ReadOptions& read_options,
                                const MultiGetRange* mget_range,
                                const SliceTransform* prefix_extractor,
                                bool skip_filters) {
-  BlockCacheLookupContext lookup_context{BlockCacheLookupCaller::kUserMGet};
+  BlockCacheLookupContext lookup_context{TableReaderCaller::kUserMGet};
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply table/block_based/block_based_table_reader.cc]: Sure. I renamed to kUserMultiGet.

File: table/table_reader_caller.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  kCompactionRefill = 11,
+  // Table reader benchmark.
+  kTableReaderBench = 12,
+  // Unit tests that call a table reader.
+  kTest = 13,
+  // After building a table, it may load all its blocks into the block cache if
+  // paranoid_file_checks is true.
+  kBuildTable = 14,
+  // sst_file_reader.
+  kUnknown = 15,
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply table/table_reader_caller.h]: I changed it to SSTFileReader.

File: table/table_reader_caller.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  kSSTDumpTool = 6,
+  kExternalSSTIngestion = 7,
+  kRepair = 8,
+  kPrefetch = 9,
+  kCompaction = 10,
+  // A compaction job may refill the block cache with blocks in the new SST
+  // files if paranoid_file_checks is true.
+  kCompactionRefill = 11,
+  // Table reader benchmark.
+  kTableReaderBench = 12,
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply table/table_reader_caller.h]: I added a kUncategorized and removed kTest and kTableReaderBench.

File: tools/block_cache_trace_analyzer_test.cc [hunk 7]
```diff
@@ -45,23 +45,23 @@ class BlockCacheTracerTest : public testing::Test {
...
-        return BlockCacheLookupCaller::kUserMGet;
+        return TableReaderCaller::kUserMGet;
       case 4:
-        return BlockCacheLookupCaller::kUserIterator;
+        return TableReaderCaller::kUserIterator;
     }
     // This cannot happend.
     assert(false);
-    return BlockCacheLookupCaller::kUserGet;
+    return TableReaderCaller::kUserGet;
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply tools/block_cache_trace_analyzer_test.cc]: Yes. I updated the code.

File: table/table_reader_caller.h (hunk 1 above)
Comment (HaoyuHuang) [reply table/table_reader_caller.h]: Sure. I added a kUncategorized enum and comments inline. "A user may use kUncategorized if the caller is not useful for analysis or the table reader is called in the test environment, e.g., unit test, table reader benchmark, etc."

**Comment by HaoyuHuang** [comment]
Thank you for your comments.

**Pull request closed without merging**